File: apps/backend/src/scheduler/pacing.ts
/**
 * Темп постановки занятий: сколько пар дисциплины должно быть поставлено
 * к концу очередной недели. Цель накопительная и пропорциональна числу
 * доступных учебных дней (короткие недели получают меньше пар). Если на неделе
 * поставить не удалось, цель следующей недели растёт — отставание компенсируется.
 */
export function effectiveRate(required: number, availableWeeks: number, weeklyRate: number | null): number {
  if (availableWeeks <= 0) return required;
  const even = required / availableWeeks;
  return Math.max(even, weeklyRate ?? 0);
}

/**
 * Накопительная цель:
 *  - равномерная часть: required × (доступных дней к концу недели / всего доступных дней);
 *  - желаемый темп (weeklyRate, пар в неделю) может «опережать» равномерный график.
 */
export function cumulativeTargetByDays(
  required: number,
  daysUpTo: number,
  totalDays: number,
  weeksUpTo: number,
  weeklyRate: number | null,
): number {
  if (totalDays <= 0 || daysUpTo >= totalDays) return required;
  const even = (required * daysUpTo) / totalDays;
  const byRate = weeklyRate ? weeklyRate * weeksUpTo : 0;
  return Math.min(required, Math.round(Math.max(even, byRate) + 1e-9));
}

File: apps/solver/app/pacing.py
"""Темп постановки занятий (совпадает с реализацией backend: apps/backend/src/scheduler/pacing.ts)."""

from __future__ import annotations


def effective_rate(required: int, available_weeks: int, weekly_rate: float | None) -> float:
    """Пар в неделю: равномерный темп или желаемый (если он выше)."""
    if available_weeks <= 0:
        return float(required)
    return max(required / available_weeks, weekly_rate or 0.0)


def cumulative_target_by_days(
    required: int,
    days_up_to: int,
    total_days: int,
    weeks_up_to: int,
    weekly_rate: float | None,
) -> int:
    """Сколько пар должно быть поставлено к концу недели.

    Цель пропорциональна числу доступных учебных дней (короткие недели получают меньше пар),
    желаемый темп weekly_rate может опережать равномерный график. Отставание компенсируется
    автоматически, так как цель накопительная.
    """
    if total_days <= 0 or days_up_to >= total_days:
        return required
    even = required * days_up_to / total_days
    by_rate = (weekly_rate or 0.0) * weeks_up_to
    # Округление «половина вверх», как Math.round в JavaScript
    return min(required, int(max(even, by_rate) + 0.5 + 1e-9))
